docs/DesignDocs/RemoteAttestationCollaterals.md

```diff
@@ -0,0 +1,457 @@
+
+#### Attester:
+An Attestation Function that creates Evidence by
+collecting, formatting and protecting (e.g., signing) Claims.  It
+presents Evidence to a Verifier using a conveyance mechanism or
+protocol.
+
+#### Verifier:
+An Attestation Function that accepts Evidence from an Attester using
+a conveyance mechanism or protocol.  It also accepts Known-Good-Values
+and Endorsments from an Asserter using a conveyance mechanism or protocol.
+It verifies the protection mechanisms, parses and appraises Evidence
+according to good-known valid (or known-invalid) Claims and Endorsments.
+It produces Attestation Results that are formatted and protected (e.g.,
+signed).  It presents Attestation Results to a Relying Party using
+a conveyance mechanism or protocol.
+
+#### Asserter:
+An Attestation Function that generates reference Claims
+about both the Attesting Computing Environment and the Attested
+Computing Environment.  The manufacturing and development
+processes are presumed to be trustworthy processes.  In other
+words the Asserter is presumed, by a Verifier, to produce valid
+Claims.  The function collects, formats and protects (e.g. signs)
+valid Claims known as Endorsements and Known-Good-Values.  It
+presents provable Claims to a Verifier using a conveyance
+mechanism or protocol.
+
+#### Relying Party:
+An Attestation Function that accepts Attestation
+Results from a Verifier using a conveyance mechanism or protocol.
+It assesses Attestation Results protections, parses and assesses
+Attestation Results according to an assessent context (Note:
+definition of the assessment context is out-of-scope).
+
+
+Motivation
+----------
+
+Currently, the existing endorsements used for Intel SGX quote verification are not exposed to the user.  This makes it difficult for the verifier to specify his/her own set of policies.  Adding these new APIs allows the verifier to specify a validation policy of his/her choosing.  Possible policies:
+1. The verifier has the option to specify its set of endorsements during verification.
+2. The verifier has the option to provide a datetime to use during verification.  This datetime specifies the date and time at which the verifier wants to do the verification.  If no datetime is provided, the datetime when the endorsements were created is used during verification.  The verifier can provide a datetime in the past, enabling auditing of the evidence and endorsements.
+
+
+User Experience
+---------------
+
+There are 2 scenarios.
+
+### 1. Verifier is provided with endorsements:
+In this scenario the attester/asserter provides the evidence and endorsements to the verifier.  The verifier is then free to use these to verify the TEE.
+
+##### Attester generates the evidence and endorsements (inside an enclave/TEE)
+```C
+...
+result = oe_get_evidence(
+    OE_EVIDENCE_FLAGS_REMOTE_ATTESTATION,
+    NULL, // custom_claims
+    0,
+    NULL, // opt_params
+    0,
+    &evidence,
+    &evidence_size,
+    &endorsements,
```

Comment:
(@dthaler for RATS insight)

Based on a reading of the RATS draft, the mapping of this to "endorsements" doesn't seem correct given the 4.3.2 section definition on Role Messages (as opposed to the intro section copied above):

> Endorsements:  Claims about immutable and implicit characteristics of
      the Attesting Computing Environment.  Typically, endorsement
      Claims are created by manufacturing or supply chain entities.

If the report/quote is the Evidence, then arguably, the only artifact in DCAP that could be construed to be the endorsement would be the PCK certificate, with the claim being "the attester with this key is recognized by the manufacturer", and that's not even part of the endorsements data structure defined here. Everything else here is actually the set of Known-Good-Values (literally, the QE ID and TCBInfo list the good values of claims like the SVN values in the evidence). 

I bring this up because if the mapping to endorsements is important, i.e. we want to support a model where the verifier can actually pull the PCK cert, then this iteration on E3 does not actually support that. It _could_ be supported in future versions of the SGX platform, and you should consider the versioning impact of this; in fact, is something we need to consider in the support of EPID attestation, where the actual endorsement comes sans all the KGV collateral on the verification side.

Reply:
The endorsement could be provided to the verifier in a way that is bundled with the evidence.   (E.g., if the PCK cert chain was provided with the quote, or inside the quote).   So just because it's in the quote/report doesn't mean it's not endorsement.   The question is whether the "it" is processor specific, or applies to all equivalent processors.  So a CRL in that sense could be in the endorsement category, as could anything else that is processor-independent.

Reply:
@dthaler, I'm not suggesting that the PCK cert (chain) is not an endorsement. I'm saying that of all the SGX attestation artifacts, it's the one thing that _best fits_ that description. I'm pointing out that if the PCK cert (chain) is an Endorsement, then Scenario 2 proposed in this design is explicitly not supported in the current SGX implementation, since it cannot be correctly retrieved by the verifier. It is a limitation of the Coffeelake platform and this design needs to address how this inconsistency in the abstraction is conveyed to a developer.

I'm also asking if the distinction between KGV and Endorsements is meaningful in the context of this design (or even in the RATS draft). Particularly since the use of RATS terminology was suggested by you, I'm unclear on why the choice of "collaterals" as a term that could encompass both KGV and Endorsements is less preferred than a use of "endorsements" that overloads an existing meaning distinct from KGV.

Reply:
> @dthaler, I'm not suggesting that the PCK cert (chain) is not an endorsement. I'm saying that of all the SGX attestation artifacts, it's the one thing that best fits that description.

Why would a CRL not fit that description?  (please educate me)

> I'm pointing out that if the PCK cert (chain) is an Endorsement, then Scenario 2 proposed in this design is explicitly not supported in the current SGX implementation, since it cannot be correctly retrieved by the verifier. It is a limitation of the Coffeelake platform and this design needs to address how this inconsistency in the abstraction is conveyed to a developer.

Thanks for the info, I'm sure there's lots of platform limitations in different platforms.   Many of the assumptions SGX has will probably not apply in general, so it's not particularly surprising that a general assumption may not apply to current SGX either.

> I'm also asking if the distinction between KGV and Endorsements is meaningful in the context of this design (or even in the RATS draft). 

The RATS discussion is still in flux.  My opinion is that KGV as a term is not a helpful term, as it implies an assumption that there is such a thing, as opposed to "security policy" for which KGV is an example, but there may be more complex logic like known good ranges, or criteria like value of claim X must be > value of claim Y, etc.   The rats arch doc is not a WG item yet because all these sorts of discussions are going on.   https://tools.ietf.org/html/draft-thaler-rats-architecture-00 is my take at writing up a proposal that some believe might be easier to understand.

> Particularly since the use of RATS terminology was suggested by you, I'm unclear on why the choice of "collaterals" as a term that could encompass both KGV and Endorsements is less preferred than a use of "endorsements" that overloads an existing meaning distinct from KGV.

If you want, I can take the suggestion (or you can) to the RATS WG to use the term "collaterals" in the way you suggest.  I just want to align between OE and RATS, and either or both might change terms.